### ISSUE TITLE
global config override

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#1940]
Provide a single configuration tiddler that declares all the overrides a user wishes to perform without ever touching shadows, possibly even real tiddlers.

This could be as declarative as some "multi-tid" format:

**$:/conig/overrides**

```
title:$:/core/ui/SideBar/Tools
tags: $:/core/ui/MoreSideBar

title:$:/DefaultTiddlers
text: MyStart
```

Starting a tiddler to be overridden with the line `title:` followed by the title and then the fields to be overwritten. This tiddler could be indexed on startup or whenever edited and applied acordingly whenever a shadow (or even a real tiddler) is loaded. As for the control-panel, a generic linked icon could indicate that a setting is actually coming from this config tiddler rather than the shadow being referenced.

This would simplify configuration, migration as well as upgrading TiddlyWiki a whole lot.

**Comment by twMat** [comment]
(hehe, the tb is back and he is causing rumble!)

Sounds like a really cool idea!
But/And, I know what the general consensus on a matter like this is; It makes sense to first make it as a plugin to try it out.

I find that I typically need to at least see the code that I'm about to overwrite. It would make sense to have some kind of transcluded view of the original code. 

Taking the idea a bit further, at least if it is a plugin, I wouldn't mind a checkbox or button to toggle between my overwrite and the core version.

**Comment by tobibeer** [comment]
It cannot be a plugin as it requires the core to handle shadows differently... by allowing a user to overwrite shadows partially, instead of entirely.

**Comment by twMat** [comment]
Hm, ok. 

Just a non-thought through idea: Woudl it be possible to fieldmange/overwrite the shadow field content with transclusions to the user content? 

```
title:$:/someshadow
tags:{{mystuff!!tags}}
```

Maybe shadow tiddlers could be defined to have this and the values we currently see are fallbacks when there is no overwrite.

>What's more, these could also be easily shared as "custom setups".

Ooh, interesting.

**Comment by Jermolene** [comment]
If I'm understanding the use-case correctly, then #1573 is the solution I favour because it is simpler, not requiring any new mechanisms. 

> Creating this as a plugin is nearly impossible as it requires the core to handle shadows differently, allowing a user to overwrite shadows partially, instead of entirely

Changing the way that shadow tiddlers work feels to me like a last-resort, something that I'd only want to change if there was a big benefit.

**Comment by pmario** [comment]
@tobibeer 

imo no new mechanism needed. It can be done out of the box already. 

most TW configurations are stored in tiddlers. So you `config/setup` just needs to be a pluginm, that overwrites those setup tiddlers. Since plugin tiddlers are shadow tiddlers there is no problem. You just need to make sure that your plugin is the last one that is executed. ....

**Comment by tobibeer** [comment]
@Jermolene, how can #1573 help to partially override shadows without modifying them?

@pmario, the key here is to **not** override entire tiddlers but only parts of them, that is fields. However, without touching the underlying tiddler.

Another way to look at it would be that shadows actually reference a kind of "hidden" namespace which could be overrun in this plugin-like form, of which I would suggest there can only be one that is active at a time so as to not overcomplicate the task of keeping track of what is actually causing a certain setup.

In other words, while a wiki may hold have a number of such setup tiddlers / configurations only one could be active at a time.

**Comment by Jermolene** [comment]
> @Jermolene, how would you envision #1573 help partially override shadows without modifying them?

It doesn't. Instead of changing the model of how shadow tiddlers work, it just provides a mechanism for assigning values to configuration tiddlers at startup.

**Comment by tobibeer** [comment]
>  it just provides a mechanism for assigning values to configuration tiddlers at startup

Mhhh, the "beef" with that is that in order to modify, say the tabs or tab order in the sidebar one has to actually touch the shadows holding the templates. With this enhancement, one could simply just override the tags field of the corresponding shadow without modifying the actual template.

**Comment by Jermolene** [comment]
It's a nice feature, but it comes at quite a cost in terms of complexity, both for the implementation and the user. 

> This tiddler could be indexed on startup and applied acordingly whenever a shadow is loaded.

It's expensive to introduce this kind of processing.

> As for the control-panel, a generic linked icon could indicate that a setting is actually coming from this setup tiddler rather than the shadow being referenced.

Again, this is an incredible amount of complexity to introduce into the control panel.

An alternative approach for upgrade-friendly overrides of shadow tiddlers is to improve support for accessing subtiddlers; the existing support already makes it possible to access the fields of an overridden shadow tiddler, if you know which plugin it came from. Being able to delegate to the original shadow tiddler is often enough to be able to withstand changes to it.

**Comment by tobibeer** [comment]
Perhaps "Setup Plugin" could actually do the job of overriding shadows by merging the current shadows with the setup overrides.

So this could mean that to modify  field of a shadow involved, one would be forced to use the corresponding setup tiddler of this plugin ...as it might otherwise overwrite your modifications.

Or is it possible for a plugin to dynamically create new shadows by merging changes into and overwriting existing ones?

**Comment by Jermolene** [comment]
> Or is it possible for a plugin to dynamically create new shadows by merging changes into and overwriting existing ones?

No, not at the moment. Shadow tiddlers are private to the wiki instance defined in boot.js, and can't be directly manipulated by external code.

Have you seen how the browser-sniff plugin works? It dynamically constructs a plugin during startup that provides values for a bunch of information about the browser. The same technique could be used to dynamically create a plugin that overrode shadow tiddlers.

**Comment by tobibeer** [comment]
> Have you seen how the browser-sniff plugin works? It dynamically constructs a plugin during startup that provides values for a bunch of information about the browser. The same technique could be used to dynamically create a plugin that overrode shadow tiddlers.

Nope. Will check.

**Comment by twMat** [comment]
(might there be a way to hijack a tiddler title? Steering any references to its fields elsewhere?)

**Comment by Infurnoape** [comment]
I just assumed that a core override in a plugin was ok. That's how I wrote http://newfieldinlinelist.tiddlyspot.com. Now, my assumption is totally shaken. Back to the drawing board again. Shadow tiddlers are so hard to figure out.

**Comment by pmario** [comment]
>the key here is to not override entire tiddlers but only parts of them, that is fields. However, without touching the underlying tiddler.

I think, that's totally confusing. We have a simple and powerful mechanism based on "the tiddler is the thing". If users want to modify core settings, they can. But they should take responsibility for their actions. Just because they install 3rd party plugins and don't clean up there broken stuff, it's __not__ the core, that is responsible if updates fail. Nor the update mechanism. 

IMO your proposal just moves the point of failure into an area where it is really hard to find and maintain.  

just my thoughts.

**Comment by tobibeer** [comment]
thing is, at some point even a tiddler is too big an entity to fiddle with, see sidebar tabs tags ...and bazillion config tiddlers ain't a sensible way to go, imho

**Comment by twMat** [comment]
> "the tiddler is the thing"

In some sense, "the *field* is the thing"; As far as I can tell, most actual communication takes place between fields in different tiddlers and not between the fields within the tiddler. As such, the tiddler merely a place holder and address to access the field.

**Comment by twMat** [comment]
@Infurnoape 
>I just assumed that a core override in a plugin was ok. 

Thing is, core updates aren't reflected then and other plugins depending on the original core tid might not behave properly. There's the risk an end user doesn't understand this or forgets the overwrite at some stage like when installing yet *another* plugin... which he must then know depends on that tiddler. That is a lot to demand from an end user.

**Comment by Infurnoape** [comment]
Please forgive me for the following tongue-in-cheek satire. After running configuration options past Charlie, he has decided that until the option to hide core components is available in a UI, I should go ahead and continue overriding shaddow tiddlers in my plugins as long as a warning is displayed clearly about what the override means in the readme and as long as I use his icon replacing the plugin icon. I still have to update my plugins with warnings. Charlie's icon may also be updated depending on UI developments. Charlie is a talking dog. Thank you.

**Comment by tobibeer** [comment]
Surely, with this kind of solution the non-updateable nature of overwritten core components would be drastically lessened, precise edits becoming more visible and, imho, more maintainable, more easily distributable.

**Comment by buggyj** [comment]
I think the problem would be better solve by moving all configuration parameters into config tiddlers, rather than using tags.
Lists could be place in data tiddlers.

**Comment by Jermolene** [comment]
@Infurnoape I don't think it's a problem for a plugin like yours to override a shadow tiddler. It's one of the ways that TiddlyWiki is designed to be customised. The user experience isn't too bad: if they want the core behaviour back they just remove the plugin.

As discussed elsewhere, there are degrees of intrusiveness/brittleness across the various ways of customising TiddlyWiki. In the core, we continue to try to support the customisations people want by trying to ensure that important customisations can be made non-intrusively.

But there are still many cases where it makes sense for a plugin to replace core tiddlers.
